modules/improvedigitalBidAdapter.js

```diff
@@ -58,130 +125,136 @@ export const spec = {
         // Google Ad Tech Provider IDs
         const atpIds = additionalConsent.substring(additionalConsent.indexOf('~') + 1);
         deepSetValue(
-          requestParameters,
+          request,
           'user.ext.consented_providers_settings.consented_providers',
           atpIds.split('.').map(id => parseInt(id, 10))
         );
       }
     }
 
+    // US Privacy
     if (bidderRequest && bidderRequest.uspConsent) {
-      requestParameters.usPrivacy = bidderRequest.uspConsent;
+      deepSetValue(request, 'regs.ext.us_privacy', bidderRequest.uspConsent);
     }
 
+    // Site Page
     if (bidderRequest && bidderRequest.refererInfo && bidderRequest.refererInfo.referer) {
-      requestParameters.referrer = bidderRequest.refererInfo.referer;
+      deepSetValue(request, 'site.page', bidderRequest.refererInfo.referer);
     }
 
-    requestParameters.schain = bidRequests[0].schain;
+    // Adding first party data
+    if (fpd) {
+      if (fpd.site) {
+        request.site = {...request.site, ...fpd.site};
+      } else if (fpd.app) {
+        request.app = fpd.app;
+      }
+      if (fpd.device) {
+        request.device = {...request.device, ...fpd.device}
+      }
+      if (getDNT()) {
+        request.device.dnt = 1;
+      }
+    }
+    // End of adding first party data
+
+    const bidRequest = bidRequests[0];
+
+    if (bidRequest.schain) {
+      deepSetValue(request, 'source.ext.schain', bidRequest.schain);
+    }
+
+    if (bidRequest.transactionId) {
+      deepSetValue(request, 'source.tid', bidRequest.transactionId);
+    }
 
-    if (bidRequests[0].userId) {
-      const eids = createEidsArray(bidRequests[0].userId);
+    if (bidRequest.userId) {
+      const eids = createEidsArray(bidRequest.userId);
       if (eids.length) {
-        deepSetValue(requestParameters, 'user.ext.eids', eids);
+        deepSetValue(request, 'user.ext.eids', eids);
       }
     }
 
-    let requestObj = idClient.createRequest(
-      normalizedBids, // requestObject
-      requestParameters
-    );
-
-    if (requestObj.errors && requestObj.errors.length > 0) {
-      logError('ID WARNING 0x01');
+    if (singleRequestMode || bidRequests.length === 1) {
```

Comment:
I can't help but I don't like how this code and the code below repeats itself with the same return object. Please refactor.